specs/Deployment/DeploymentAPI.md

```diff
@@ -0,0 +1,141 @@
+1. Does this system have the min version of the Windows App SDK I require and are they usable?
+2. What do I do to resolve this if it doesn't?
+
+The API addresses #1 by a simple query of the version they seek and a possible check for OS
+updates. #2 has two possible answers. First is direct resolution by the developer by acquiring and
+installing the packages if possible. Since the main packages are small and the target developer
+scenario will have permissions to deploy them, we will minimize developer and user pain by
+including the main package MSIX packages within the framework, and this API will allow for the
+deployment of that package if needed.
+
+Processes lacking permission to deploy MSIX are out-of-scope at this time.
+
+# Examples
+
+## GetStatus
+
+This is the means by which an app or program can call to verify all required packages are present
+to a minimum version needed by the loaded framework.  A process should call GetStatus() after
+process initialization and before using Windows App Runtime content (e.g. in main() or WinMain()).
+
+Developers call this method without parameters, as all information about the version, channel, and
+architecture are derived from the current WARfwk loaded. This method is good to identify if there
+is possible work required in the form of package install which may take several seconds to
+complete.
+
+```C++
+auto status{Microsoft::Windows::ApplicationModel::WindowsAppRuntime::Deployment::GetStatus()};
+if (!status.IsOK())
+{
+    ...
+}
+```
+
+## Initialize
+
+This API performs a status check and if any packages are required for install, attempts to register
+those packages and get the WindowsAppRuntime into an OK state. This will perform package install if
+possible. It is possible multiple packages may be required for install, and this will attempt
+to deploy all required packages. The returned WindowsAppRuntimeStatus will contain the first error
+if one occurs during package install. There will be no rollback of successfully installed packages
+if an error occurs on a subsequent package install.
+
+
+```C++
+auto status{Microsoft::Windows::ApplicationModel::WindowsAppRuntime::Deployment::Initialize()};
+if (!status.IsOK())
+{
+    return status.ExtendedError();
+}
+```
+
+
+# API Details
+
+```c#
+namespace Microsoft.Windows.ApplicationModel.WindowsAppSDK
+{
+    /// Represents the current Deployment status of the WindowsAppRuntime
+    [experimental]
+    runtimeclass DeploymentStatus
+    {
+        DeploymentStatus(
+            Boolean isOK,
+            Boolean packageInstallRequired,
+            HRESULT extendedError);
+
+        /// Returns True if WindowsAppRuntime has all required packages at equal or greater version
+        /// than the current framework requires.
+        Boolean IsOK{ get; };
+
+        /// Returns True if a Package install is required the WindowsAppRuntime to be in a good
+        /// state.
+        Boolean PackageInstallRequired{ get; };
```

Comment:
So the intent is the developer can know if the main (singleton, things like WNS) are present and accounted for, and they would do this by calling the API maybe twice, once with the main and once with the FXP?

Reply:
The intent is so the developer can know whether the reason the runtime is in a not-OK state is due to missing packages, which calling Initialize() will attempt to resolve. So the idea is that GetStatus() is a performant check, while Initialize() can also do package installs if the status is not OK and may not be instantaneous. I think the PackageInstallRequired boolean is not particularly useful since the action a developer would take (if GetStatus() is not ok, call Initialize()) is essentially the same, since Initialize() will take whatever actions are possible to resolve the not-OK state and turn it into an OK state.

The API could return a list of packages that are missing, but I don't think this is particularly actionable, since the framework contains the missing packages and calling Initialize() will attempt to deploy them if any are missing. However Initialize() could fail and for diagnostic purposes the list of missing packages may be useful, but that can also be done with telemetry.